File: clients/tia_hooks.go
/*
TiaCC Go Client - 测试框架集成钩子

Usage:

	hooks := tia.NewHooks(nil)
	if err := hooks.Connect(); err != nil {
	    log.Fatal(err)
	}
	defer hooks.Disconnect()

	for _, test := range tests {
	    hooks.BeforeTest(test.Name)
	    runTest(test)
	    hooks.AfterTest(test.Name)
	}
*/
package tia

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net"
	"sync"
	"time"
)

type Config struct {
	Host       string
	Port       int
	Timeout    time.Duration
	Mode       string // "precise" or "bucket"
	BucketSize int
	Language   string // "cpp" or "csharp"
}

func DefaultConfig() *Config {
	return &Config{
		Host:       "127.0.0.1",
		Port:       19840,
		Timeout:    5 * time.Second,
		Mode:       "precise",
		BucketSize: 50,
		Language:   "cpp",
	}
}

type Hooks struct {
	config      *Config
	conn        net.Conn
	reader      *bufio.Reader
	requestID   int
	bucketCount int
	bucketTests []string
	mu          sync.Mutex
}

func NewHooks(config *Config) *Hooks {
	if config == nil {
		config = DefaultConfig()
	}
	return &Hooks{config: config}
}

func (h *Hooks) Connect() error {
	addr := fmt.Sprintf("%s:%d", h.config.Host, h.config.Port)
	conn, err := net.DialTimeout("tcp", addr, h.config.Timeout)
	if err != nil {
		return fmt.Errorf("连接失败: %w", err)
	}

	h.conn = conn
	h.reader = bufio.NewReader(conn)
	return nil
}

func (h *Hooks) Disconnect() {
	if h.conn != nil {
		h.conn.Close()
		h.conn = nil
	}
}

type rpcRequest struct {
	JSONRPC string      `json:"jsonrpc"`
	Method  string      `json:"method"`
	Params  interface{} `json:"params"`
	ID      int         `json:"id"`
}

type rpcResponse struct {
	JSONRPC string          `json:"jsonrpc"`
	Result  json.RawMessage `json:"result"`
	Error   *rpcError       `json:"error"`
	ID      int             `json:"id"`
}

type rpcError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

func (h *Hooks) sendRPC(method string, params interface{}) (json.RawMessage, error) {
	h.mu.Lock()
	defer h.mu.Unlock()

	if h.conn == nil {
		return nil, fmt.Errorf("未连接")
	}

	h.requestID++
	req := rpcRequest{
		JSONRPC: "2.0",
		Method:  method,
		Params:  params,
		ID:      h.requestID,
	}

	data, _ := json.Marshal(req)
	data = append(data, '\n')

	h.conn.SetDeadline(time.Now().Add(h.config.Timeout))
	if _, err := h.conn.Write(data); err != nil {
		return nil, err
	}

	line, err := h.reader.ReadBytes('\n')
	if err != nil {
		return nil, err
	}

	var resp rpcResponse
	if err := json.Unmarshal(line, &resp); err != nil {
		return nil, err
	}

	if resp.Error != nil {
		return nil, fmt.Errorf("RPC 错误: %s", resp.Error.Message)
	}

	return resp.Result, nil
}

func (h *Hooks) StartRecording(testID string) (bool, error) {
	result, err := h.sendRPC("startRecording", map[string]string{
		"testId":   testID,
		"language": h.config.Language,
	})
	if err != nil {
		return false, err
	}

	var res struct{ Success bool }
	json.Unmarshal(result, &res)
	return res.Success, nil
}

func (h *Hooks) StopRecording(testID string) (bool, error) {
	result, err := h.sendRPC("stopRecording", map[string]string{"testId": testID})
	if err != nil {
		return false, err
	}

	var res struct{ Success bool }
	json.Unmarshal(result, &res)
	return res.Success, nil
}

func (h *Hooks) DumpCoverage(testID, outputPath string) (bool, error) {
	params := map[string]string{"testId": testID}
	if outputPath != "" {
		params["outputPath"] = outputPath
	}

	result, err := h.sendRPC("dumpCoverage", params)
	if err != nil {
		return false, err
	}

	var res struct{ Success bool }
	json.Unmarshal(result, &res)
	return res.Success, nil
}

func (h *Hooks) ResetAll() (bool, error) {
	result, err := h.sendRPC("resetAll", nil)
	if err != nil {
		return false, err
	}

	var res struct{ Success bool }
	json.Unmarshal(result, &res)
	return res.Success, nil
}

// 高级 API

func (h *Hooks) BeforeTest(testID string) error {
	if h.config.Mode == "precise" {
		_, err := h.StartRecording(testID)
		return err
	}

	h.bucketTests = append(h.bucketTests, testID)
	h.bucketCount++

	if h.bucketCount == 1 {
		bucketID := fmt.Sprintf("bucket_%d", h.requestID/h.config.BucketSize)
		_, err := h.StartRecording(bucketID)
		return err
	}
	return nil
}

func (h *Hooks) AfterTest(testID string) error {
	if h.config.Mode == "precise" {
		h.StopRecording(testID)
		_, err := h.DumpCoverage(testID, "")
		return err
	}

	if h.bucketCount >= h.config.BucketSize {
		return h.FlushBucket()
	}
	return nil
}

func (h *Hooks) FlushBucket() error {
	if h.bucketCount > 0 {
		bucketID := fmt.Sprintf("bucket_%d", h.requestID/h.config.BucketSize)
		h.StopRecording(bucketID)
		h.DumpCoverage(bucketID, "")
		h.bucketTests = nil
		h.bucketCount = 0
	}
	return nil
}

File: clients/tia_hooks.ts
/**
 * TiaCC TypeScript/JavaScript Client - 测试框架集成钩子
 *
 * Usage:
 *     import { TiaHooks } from './tia_hooks';
 *
 *     const hooks = new TiaHooks();
 *     await hooks.connect();
 *
 *     for (const test of tests) {
 *         await hooks.beforeTest(test.name);
 *         await runTest(test);
 *         await hooks.afterTest(test.name);
 *     }
 *
 *     hooks.disconnect();
 */

import { Socket } from 'net';

export interface TiaConfig {
  host?: string;
  port?: number;
  timeout?: number;
  mode?: 'precise' | 'bucket';
  bucketSize?: number;
  language?: 'cpp' | 'csharp';
}

export class TiaHooks {
  private config: Required<TiaConfig>;
  private socket: Socket | null = null;
  private requestId = 0;
  private bucketCount = 0;
  private bucketTests: string[] = [];
  private responseBuffer = '';

  constructor(config: TiaConfig = {}) {
    this.config = {
      host: config.host ?? '127.0.0.1',
      port: config.port ?? 19840,
      timeout: config.timeout ?? 5000,
      mode: config.mode ?? 'precise',
      bucketSize: config.bucketSize ?? 50,
      language: config.language ?? 'cpp',
    };
  }

  connect(): Promise<boolean> {
    return new Promise((resolve) => {
      this.socket = new Socket();
      this.socket.setTimeout(this.config.timeout);

      this.socket.on('error', (err) => {
        console.error(`[TiaCC] 连接错误: ${err.message}`);
        resolve(false);
      });

      this.socket.connect(this.config.port, this.config.host, () => {
        resolve(true);
      });
    });
  }

  disconnect(): void {
    this.socket?.destroy();
    this.socket = null;
  }

  private sendRpc(method: string, params?: Record<string, unknown>): Promise<any> {
    return new Promise((resolve, reject) => {
      if (!this.socket) {
        reject(new Error('Not connected'));
        return;
      }

      const request = {
        jsonrpc: '2.0',
        method,
        params: params ?? {},
        id: ++this.requestId,
      };

      const onData = (data: Buffer) => {
        this.responseBuffer += data.toString();
        const newlineIndex = this.responseBuffer.indexOf('\n');

        if (newlineIndex !== -1) {
          const line = this.responseBuffer.slice(0, newlineIndex);
          this.responseBuffer = this.responseBuffer.slice(newlineIndex + 1);

          this.socket?.off('data', onData);

          try {
            const response = JSON.parse(line);
            resolve(response.result);
          } catch (e) {
            reject(e);
          }
        }
      };

      this.socket.on('data', onData);
      this.socket.write(JSON.stringify(request) + '\n');
    });
  }

  async startRecording(testId: string): Promise<boolean> {
    const result = await this.sendRpc('startRecording', {
      testId,
      language: this.config.language,
    });
    return result?.success ?? false;
  }

  async stopRecording(testId: string): Promise<boolean> {
    const result = await this.sendRpc('stopRecording', { testId });
    return result?.success ?? false;
  }

  async dumpCoverage(testId: string, outputPath?: string): Promise<boolean> {
    const params: Record<string, string> = { testId };
    if (outputPath) params.outputPath = outputPath;

    const result = await this.sendRpc('dumpCoverage', params);
    return result?.success ?? false;
  }

  async resetAll(): Promise<boolean> {
    const result = await this.sendRpc('resetAll');
    return result?.success ?? false;
  }

  async getStatus(): Promise<{ recording: boolean; testId: string; runtimeAvailable: boolean } | null> {
    return this.sendRpc('getStatus');
  }

  // 高级 API
  async beforeTest(testId: string): Promise<void> {
    if (this.config.mode === 'precise') {
      await this.startRecording(testId);
    } else {
      this.bucketTests.push(testId);
      this.bucketCount++;
      if (this.bucketCount === 1) {
        await this.startRecording(`bucket_${Math.floor(this.requestId / this.config.bucketSize)}`);
      }
    }
  }

  async afterTest(testId: string): Promise<void> {
    if (this.config.mode === 'precise') {
      await this.stopRecording(testId);
      await this.dumpCoverage(testId);
    } else if (this.bucketCount >= this.config.bucketSize) {
      await this.flushBucket();
    }
  }

  async flushBucket(): Promise<void> {
    if (this.bucketCount > 0) {
      const bucketId = `bucket_${Math.floor(this.requestId / this.config.bucketSize)}`;
      await this.stopRecording(bucketId);
      await this.dumpCoverage(bucketId);
      this.bucketTests = [];
      this.bucketCount = 0;
    }
  }
}

// 使用示例
async function main() {
  const hooks = new TiaHooks({ mode: 'precise' });

  if (await hooks.connect()) {
    console.log('状态:', await hooks.getStatus());

    await hooks.beforeTest('test_example');
    // ... 运行测试 ...
    await hooks.afterTest('test_example');

    hooks.disconnect();
  }
}
